`JsonPointer.head()` throws `StringIndexOutOfBoundsException` for valid JSON Pointers

**Problem Description:**
After upgrading to Jackson 2.19.0, valid `JsonPointer` strings are causing `StringIndexOutOfBoundsException` when calling `head()`.
The crash happens due to a negative `suffixLength` calculated in `_constructHead()`.
This was not an issue in previous versions (2.18.x and earlier).

```
Caused by: java.lang.StringIndexOutOfBoundsException: begin 0, end 14, length 6
    at java.base/java.lang.String.checkBoundsBeginEnd(String.java:4606)
    at java.base/java.lang.String.substring(String.java:2709)
    at com.fasterxml.jackson.core.JsonPointer._constructHead(JsonPointer.java:853)
    at com.fasterxml.jackson.core.JsonPointer.head(JsonPointer.java:606)
```

**Reproducer:**

```java
import com.fasterxml.jackson.core.JsonPointer;

public class JsonPointerBugTest {
    public static void main(String[] args) {
        JsonPointer head = JsonPointer.compile("/a/b/0/qwerty");
        head = head.head();
        head = head.head(); // Exception happens here
        System.out.println(head);
    }
}
```

**Version:**
Jackson Core 2.19.0

--- a/release-notes/VERSION-2.x
+++ b/release-notes/VERSION-2.x
@@ -14,6 +14,12 @@ a pure JSON library.
 === Releases ===
 ------------------------------------------------------------------------
 
+2.19.1 (not yet relesed)
+
+#1425: `JsonPointer.head()` throws `StringIndexOutOfBoundsException` for
+  valid JSON Pointers
+ (reported by @sergeykad)
+
 2.19.0 (24-Apr-2025)
 
 #633: Allow skipping `RS` CTRL-CHAR to support JSON Text Sequences

--- a/src/main/java/com/fasterxml/jackson/core/JsonPointer.java
+++ b/src/main/java/com/fasterxml/jackson/core/JsonPointer.java
@@ -169,7 +169,15 @@ public class JsonPointer implements Serializable
         _matchingElementIndex = matchIndex;
     }
 
-    // @since 2.19
+    /**
+     * Copy-constructor used for creating transformed instances with
+     * re-linking textual contents to new "next" pointer instance.
+     *
+     * @param src Original pointer to copy "full String" from
+     * @param next New "next" pointed to link to
+     *
+     * @since 2.19
+     */
     protected JsonPointer(JsonPointer src, JsonPointer next) {
         _asString = src._asString;
         _asStringOffset = src._asStringOffset;
@@ -178,7 +186,16 @@ public class JsonPointer implements Serializable
         _matchingElementIndex = src._matchingElementIndex;
     }
 
-    // @since 2.19
+    /**
+     * Copy-constructor used for creating transformed instances without
+     * "next" linkage
+     *
+     * @param src Original pointer to copy "matchingXxx" fields from
+     * @param newFullString Full String to use
+     * @param newFullStringOffset Offset for new full String to use
+     *
+     * @since 2.19
+     */
     protected JsonPointer(JsonPointer src, String newFullString, int newFullStringOffset) {
         _asString = newFullString;
         _asStringOffset = newFullStringOffset;
@@ -841,22 +858,22 @@ public class JsonPointer implements Serializable
         if (last == this) {
             return EMPTY;
         }
-        // and from that, length of suffix to drop
-        final int suffixLength = last.length();
 
         // Initialize a list to store intermediate JsonPointers in reverse
         ArrayList<JsonPointer> pointers = new ArrayList<>();
 
         JsonPointer current = this;
-        String fullString = toString();
+        String origFullString = toString();
         // Make sure to share the new full string for path segments
-        fullString = fullString.substring(0, fullString.length() - suffixLength);
-        int offset = 0;
+        String fullString = origFullString.substring(0, origFullString.length() - last.length());
+
+        // Also: if there was an offset, must compensate (new String starts at 0)
+        final int offsetDiff = -_asStringOffset;
 
         while (current != last) {
+            // NOTE: since we drop from the end we can simply reuse offset (w/ possible modification)
             JsonPointer nextSegment = new JsonPointer(current,
-                    fullString, offset);
-            offset += suffixLength;
+                    fullString, current._asStringOffset + offsetDiff);
             pointers.add(nextSegment);
             current = current._nextSegment;
         }

--- a/src/test/java/com/fasterxml/jackson/core/jsonptr/JsonPointer1425Test.java
+++ b/src/test/java/com/fasterxml/jackson/core/jsonptr/JsonPointer1425Test.java
@@ -1,0 +1,26 @@
+package com.fasterxml.jackson.core.jsonptr;
+
+import org.junit.jupiter.api.Test;
+
+import com.fasterxml.jackson.core.JUnit5TestBase;
+import com.fasterxml.jackson.core.JsonPointer;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class JsonPointer1425Test extends JUnit5TestBase
+{
+
+    // [core#1425]
+    @Test
+    public void test1425() {
+        JsonPointer ptr = JsonPointer.compile("/a/b/0/qwerty");
+        JsonPointer head = ptr.head();
+        assertEquals("/a/b/0", head.toString());
+        head = head.head(); // Exception happens here
+        assertEquals("/a/b", head.toString());
+        head = head.head();
+        assertEquals("/a", head.toString());
+        head = head.head();
+        assertEquals("", head.toString());
+    }
+}

--- a/src/test/java/com/fasterxml/jackson/core/jsonptr/JsonPointer1425Test.java
+++ b/src/test/java/com/fasterxml/jackson/core/jsonptr/JsonPointer1425Test.java
@@ -12,7 +12,7 @@ public class JsonPointer1425Test extends JUnit5TestBase
 
     // [core#1425]
     @Test
-    public void test1425() {
+    public void test1425Basic() {
         JsonPointer ptr = JsonPointer.compile("/a/b/0/qwerty");
         JsonPointer head = ptr.head();
         assertEquals("/a/b/0", head.toString());
@@ -20,6 +20,21 @@ public class JsonPointer1425Test extends JUnit5TestBase
         assertEquals("/a/b", head.toString());
         head = head.head();
         assertEquals("/a", head.toString());
+        head = head.head();
+        assertEquals("", head.toString());
+    }
+
+    // [core#1425]
+    @Test
+    public void test1425Variations() {
+        JsonPointer ptr = JsonPointer.compile("/a/b/0/qwerty");
+        JsonPointer tail = ptr.tail();
+
+        assertEquals("/b/0/qwerty", tail.toString());
+        JsonPointer head = tail.head();
+        assertEquals("/b/0", head.toString());
+        head = head.head();
+        assertEquals("/b", head.toString());
         head = head.head();
         assertEquals("", head.toString());
     }

